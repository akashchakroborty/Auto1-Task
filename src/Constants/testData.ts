import { CarsResponse } from '../redux/types/carsTypes';

export const MOCK_CARS_DATA: CarsResponse = {
  cars: [
    {
      stockNumber: 97074,
      manufacturerName: 'Porsche',
      modelName: '718 Cayman',
      color: 'black',
      mileage: { number: 100056, unit: 'km' },
      fuelType: 'Diesel',
      pictureUrl: 'https://auto1-js-task-api--mufasa71.repl.co/images/car.svg',
    },
    {
      stockNumber: 12876,
      manufacturerName: 'Tesla',
      modelName: 'Roadster',
      color: 'yellow',
      mileage: { number: 100481, unit: 'km' },
      fuelType: 'Petrol',
      pictureUrl: 'https://auto1-js-task-api--mufasa71.repl.co/images/car.svg',
    },
    {
      stockNumber: 18029,
      manufacturerName: 'Fiat',
      modelName: '131',
      color: 'white',
      mileage: { number: 100710, unit: 'km' },
      fuelType: 'Petrol',
      pictureUrl: 'https://auto1-js-task-api--mufasa71.repl.co/images/car.svg',
    },
    {
      stockNumber: 65672,
      manufacturerName: 'Skoda',
      modelName: 'Felicia',
      color: 'black',
      mileage: { number: 100751, unit: 'km' },
      fuelType: 'Petrol',
      pictureUrl: 'https://auto1-js-task-api--mufasa71.repl.co/images/car.svg',
    },
    {
      stockNumber: 42578,
      manufacturerName: 'Mercedes-Benz',
      modelName: 'Sprinter',
      color: 'yellow',
      mileage: { number: 100782, unit: 'km' },
      fuelType: 'Diesel',
      pictureUrl: 'https://auto1-js-task-api--mufasa71.repl.co/images/car.svg',
    },
    {
      stockNumber: 70917,
      manufacturerName: 'Tesla',
      modelName: 'Roadster',
      color: 'green',
      mileage: { number: 100929, unit: 'km' },
      fuelType: 'Diesel',
      pictureUrl: 'https://auto1-js-task-api--mufasa71.repl.co/images/car.svg',
    },
    {
      stockNumber: 53029,
      manufacturerName: 'Tesla',
      modelName: 'Model X',
      color: 'black',
      mileage: { number: 100997, unit: 'km' },
      fuelType: 'Diesel',
      pictureUrl: 'https://auto1-js-task-api--mufasa71.repl.co/images/car.svg',
    },
    {
      stockNumber: 77489,
      manufacturerName: 'Chrysler',
      modelName: 'Viper',
      color: 'red',
      mileage: { number: 101001, unit: 'km' },
      fuelType: 'Petrol',
      pictureUrl: 'https://auto1-js-task-api--mufasa71.repl.co/images/car.svg',
    },
    {
      stockNumber: 98739,
      manufacturerName: 'Tesla',
      modelName: 'Roadster',
      color: 'blue',
      mileage: { number: 101066, unit: 'km' },
      fuelType: 'Petrol',
      pictureUrl: 'https://auto1-js-task-api--mufasa71.repl.co/images/car.svg',
    },
    {
      stockNumber: 72351,
      manufacturerName: 'Porsche',
      modelName: '968',
      color: 'green',
      mileage: { number: 101069, unit: 'km' },
      fuelType: 'Diesel',
      pictureUrl: 'https://auto1-js-task-api--mufasa71.repl.co/images/car.svg',
    },
  ],
  totalPageCount: 100,
  totalCarsCount: 1000,
};

export const MOCK_COLORS = {
  colors: ['red', 'blue', 'green', 'black', 'yellow', 'white', 'silver'],
};

export const MOCK_MANUFACTURERS = {
  manufacturers: [
    {
      name: 'Audi',
      models: [
        {
          name: '100',
        },
      ],
    },
  ],
};
