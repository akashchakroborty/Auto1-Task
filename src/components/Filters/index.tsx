import React, { useEffect } from 'react';
import {
 Button, FormControl, MenuItem, Select, InputLabel
} from '@mui/material';
import { makeStyles } from '@mui/styles';
import { CarsRequest } from '../../redux/types/carsTypes';
import { Manufacturer } from '../../redux/types/filtersTypes';

export interface FiltersProps {
  getCars: (props: CarsRequest) => void;
  setFilters: (props: CarsRequest) => void;
  getColors: () => void;
  getManufacturers: () => void;
  manufacturers: Manufacturer[];
  colors: string[];
  activeFilters: CarsRequest;
}

const useStyles = makeStyles({
  form: {
    display: 'flex',
    flexDirection: 'column',
    margin: 24,
  },
  formControl: {
    margin: '8px 0 24px !important',
    minWidth: 120,
  },
  option: {
    textTransform: 'capitalize',
  },
  filterButton: {
    alignSelf: 'flex-end',
  },
});

const Filters: React.FC<FiltersProps> = ({
  colors,
  manufacturers,
  activeFilters,
  getManufacturers,
  setFilters,
  getCars,
  getColors,
}) => {
  const classes = useStyles();
  const { manufacturer, color, sort } = activeFilters;
  const handleColorChange = (event: { target: { value: string } }) =>
    setFilters({ color: event.target.value });

  const handleManufacturerChange = (event: { target: { value: any } }) =>
    setFilters({ manufacturer: event.target.value });

  const handleSubmit = (e: React.FormEvent) => {
    e.preventDefault();
    setFilters({ page: 1 });
    getCars({
 color, manufacturer, page: 1, sort
});
  };
  useEffect(() => {
    if (!colors.length) {
      getColors();
    }
    if (!manufacturers.length) {
      getManufacturers();
    }
  }, [getColors, getManufacturers]);
  return (
    <form className={classes.form} onSubmit={handleSubmit}>
      <InputLabel id="colorFilterLabel">Color</InputLabel>
      <FormControl variant="outlined" className={classes.formControl}>
        <Select
          id="colorFilterSelect"
          labelId="colorFilterLabel"
          value={color || ''}
          onChange={handleColorChange}
          displayEmpty
          className={classes.option}
        >
          <MenuItem value="">
            <span>All Car Colors</span>
          </MenuItem>
          {colors.map((colorOption) => (
            <MenuItem key={colorOption} value={colorOption} className={classes.option}>
              {colorOption}
            </MenuItem>
          ))}
        </Select>
      </FormControl>
      <InputLabel id="manufacturerFilterLabel">Manufacturer</InputLabel>
      <FormControl variant="outlined" className={classes.formControl}>
        <Select
          id="manufacturerFilterSelect"
          labelId="manufacturerFilterLabel"
          value={manufacturer || ''}
          onChange={handleManufacturerChange}
          displayEmpty
          className={classes.option}
        >
          <MenuItem value="">
            <span>All Manufacturers</span>
          </MenuItem>
          {manufacturers.map((manufacturerOption) => (
            <MenuItem
              key={manufacturerOption.name}
              value={manufacturerOption.name}
              className={classes.option}
            >
              {manufacturerOption.name}
            </MenuItem>
          ))}
        </Select>
      </FormControl>
      <Button
        type="submit"
        variant="contained"
        color="primary"
        fullWidth={false}
        className={classes.filterButton}
      >
        Filter
      </Button>
    </form>
  );
};

export default Filters;
