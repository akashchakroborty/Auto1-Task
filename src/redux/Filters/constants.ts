export const GET_COLORS = 'GET_COLORS';
export const GET_MANUFACTURERS = 'GET_MANUFACTURERS';
export const SET_FILTERS = 'SET_FILTERS';
