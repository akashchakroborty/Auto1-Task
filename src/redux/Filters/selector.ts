import { AppState } from '../store';
import { CarsRequest } from '../types/carsTypes';

export const selectActiveFilters = (state: AppState): CarsRequest => state.filters.activeFilters;
