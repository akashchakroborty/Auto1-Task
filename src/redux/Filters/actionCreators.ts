import { ThunkAction, ThunkDispatch } from 'redux-thunk';
import { HttpMethods, RestClient, StatusCode } from '../../utillity/restClient';
import { getColorsUrl, getManufacturersUrl } from '../../utillity/url-utils/urlBuilder';
import { AppState } from '../store';
import { CarsRequest } from '../types/carsTypes';
import { FiltersActionTypes, SetFilters } from '../types/filtersActionTypes';
import { GET_COLORS, GET_MANUFACTURERS, SET_FILTERS } from './constants';

export const getColors = (): ThunkAction<Promise<void>, AppState, void, FiltersActionTypes> =>
  async (dispatch: ThunkDispatch<AppState, void, FiltersActionTypes>) => {
    const url = getColorsUrl();
    const response = await RestClient({
      url,
      method: HttpMethods.get,
    });
    if (response.status === StatusCode.success) {
      dispatch({
        type: GET_COLORS,
        payload: response.data,
      });
    }
  };

export const getManufacturers = (): ThunkAction<Promise<void>, AppState, void, FiltersActionTypes> =>
  async (dispatch: ThunkDispatch<AppState, void, FiltersActionTypes>) => {
    const url = getManufacturersUrl();
    const response = await RestClient({
      url,
      method: HttpMethods.get,
    });
    if (response.status === StatusCode.success) {
      dispatch({
        type: GET_MANUFACTURERS,
        payload: response.data,
      });
    }
  };

export const setFilters = (props: CarsRequest): SetFilters => {
  return {
    type: SET_FILTERS,
    payload: props,
  };
};
