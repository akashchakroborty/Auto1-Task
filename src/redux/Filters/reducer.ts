import { FiltersActionTypes } from '../types/filtersActionTypes';
import { FiltersState } from '../types/filtersTypes';
import { GET_COLORS, GET_MANUFACTURERS, SET_FILTERS } from './constants';

export const defaultState: FiltersState = {
  colors: [],
  manufacturers: [],
  activeFilters: {
    page: 1,
    sort: 'asc',
  },
};

const filtersReducer = (state = defaultState, action: FiltersActionTypes): FiltersState => {
  switch (action.type) {
    case GET_COLORS:
      return {
        ...state,
        colors: action.payload,
      };
    case GET_MANUFACTURERS:
      return {
        ...state,
        manufacturers: action.payload,
      };
    case SET_FILTERS:
      return {
        ...state,
        activeFilters: { ...state.activeFilters, ...action.payload },
      };
    default:
      return state;
  }
};

export default filtersReducer;
