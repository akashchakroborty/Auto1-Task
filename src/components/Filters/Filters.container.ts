import { connect } from 'react-redux';
import { ThunkDispatch } from 'redux-thunk';
import { getCars } from '../../redux/Cars/actionCreators';
import { AppState } from '../../Redux/store';
import { CarsActionTypes } from '../../redux/types/carsActionTypes';
import { CarsRequest } from '../../redux/types/carsTypes';
import Filters from '.';
import {
  selectActiveFilters,
  selectColors,
  selectManufacturers,
} from '../../redux/Filters/selector';
import { getColors, getManufacturers, setFilters } from '../../redux/Filters/actionCreators';
import { FiltersActionTypes } from '../../redux/types/filtersActionTypes';

const mapStateToProps = (state: AppState) => ({
  colors: selectColors(state),
  manufacturers: selectManufacturers(state),
  activeFilters: selectActiveFilters(state),
});

const mapDispatchToProps = (
  dispatch: ThunkDispatch<AppState, void, CarsActionTypes | FiltersActionTypes>,
) => {
  return {
    getCars: (props: CarsRequest) => {
      // eslint-disable-next-line @typescript-eslint/no-floating-promises
      dispatch(getCars(props));
    },
    getColors: () => {
      // eslint-disable-next-line @typescript-eslint/no-floating-promises
      dispatch(getColors());
    },
    getManufacturers: () => {
      // eslint-disable-next-line @typescript-eslint/no-floating-promises
      dispatch(getManufacturers());
    },
    setFilters: (props: CarsRequest) => {
      dispatch(setFilters(props));
    },
  };
};

export default connect(mapStateToProps, mapDispatchToProps)(Filters);
