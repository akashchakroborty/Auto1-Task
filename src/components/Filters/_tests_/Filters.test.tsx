import React from 'react';
import { shallow } from 'enzyme';
import Filters, { FiltersProps } from '..';
import { MOCK_COLORS, MOCK_MANUFACTURERS } from '../../../Constants/testData';

describe('should test the Filters Component', () => {
  const mockProps: FiltersProps = {
    getCars: jest.fn(),
    setFilters: jest.fn(),
    getColors: jest.fn(),
    getManufacturers: jest.fn(),
    manufacturers: MOCK_MANUFACTURERS.manufacturers,
    colors: MOCK_COLORS.colors,
    activeFilters: { page: 1, sort: 'asc' },
  };
  beforeEach(() => {
    jest.spyOn(React, 'useEffect').mockImplementation((f) => f());
  });
  test('should render the component', () => {
    const wrapper = shallow(<Filters {...mockProps} />);
    expect(wrapper).toMatchSnapshot();
  });
  test('should test handleColorChange function', () => {
    const wrapper = shallow(<Filters {...mockProps} />);
    wrapper.find('#colorFilterSelect').simulate('change', { target: { value: 'pink' } });
    expect(mockProps.setFilters).toHaveBeenCalledWith({ color: 'pink' });
  });
  test('should test handleManufacturerChange function', () => {
    const wrapper = shallow(<Filters {...mockProps} />);
    wrapper.find('#manufacturerFilterSelect').simulate('change', { target: { value: 'Audi' } });
    expect(mockProps.setFilters).toHaveBeenCalledWith({ manufacturer: 'Audi' });
  });
  test('should test handleSubmit function', () => {
    const wrapper = shallow(<Filters {...mockProps} />);
    wrapper.find('#filterButton').simulate('click', { preventDefault: jest.fn() });
    expect(mockProps.setFilters).toHaveBeenCalledWith({ page: 1 });
    expect(mockProps.getCars).toHaveBeenCalled();
  });
});
