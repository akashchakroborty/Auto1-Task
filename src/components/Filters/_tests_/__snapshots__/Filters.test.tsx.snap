// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should test the Filters Component should render the component 1`] = `
<form
  className="makeStyles-form-1"
>
  <ForwardRef(InputLabel)
    id="colorFilterLabel"
  >
    Color
  </ForwardRef(InputLabel)>
  <ForwardRef(FormControl)
    className="makeStyles-formControl-2"
    variant="outlined"
  >
    <ForwardRef(Select)
      className="makeStyles-option-3"
      displayEmpty={true}
      id="colorFilterSelect"
      labelId="colorFilterLabel"
      onChange={[Function]}
      value=""
    >
      <ForwardRef(MenuItem)
        value=""
      >
        <span>
          All Car Colors
        </span>
      </ForwardRef(MenuItem)>
      <ForwardRef(MenuItem)
        className="makeStyles-option-3"
        key="red"
        value="red"
      >
        red
      </ForwardRef(MenuItem)>
      <ForwardRef(MenuItem)
        className="makeStyles-option-3"
        key="blue"
        value="blue"
      >
        blue
      </ForwardRef(MenuItem)>
      <ForwardRef(MenuItem)
        className="makeStyles-option-3"
        key="green"
        value="green"
      >
        green
      </ForwardRef(MenuItem)>
      <ForwardRef(MenuItem)
        className="makeStyles-option-3"
        key="black"
        value="black"
      >
        black
      </ForwardRef(MenuItem)>
      <ForwardRef(MenuItem)
        className="makeStyles-option-3"
        key="yellow"
        value="yellow"
      >
        yellow
      </ForwardRef(MenuItem)>
      <ForwardRef(MenuItem)
        className="makeStyles-option-3"
        key="white"
        value="white"
      >
        white
      </ForwardRef(MenuItem)>
      <ForwardRef(MenuItem)
        className="makeStyles-option-3"
        key="silver"
        value="silver"
      >
        silver
      </ForwardRef(MenuItem)>
    </ForwardRef(Select)>
  </ForwardRef(FormControl)>
  <ForwardRef(InputLabel)
    id="manufacturerFilterLabel"
  >
    Manufacturer
  </ForwardRef(InputLabel)>
  <ForwardRef(FormControl)
    className="makeStyles-formControl-2"
    variant="outlined"
  >
    <ForwardRef(Select)
      className="makeStyles-option-3"
      displayEmpty={true}
      id="manufacturerFilterSelect"
      labelId="manufacturerFilterLabel"
      onChange={[Function]}
      value=""
    >
      <ForwardRef(MenuItem)
        value=""
      >
        <span>
          All Manufacturers
        </span>
      </ForwardRef(MenuItem)>
      <ForwardRef(MenuItem)
        className="makeStyles-option-3"
        key="Audi"
        value="Audi"
      >
        Audi
      </ForwardRef(MenuItem)>
    </ForwardRef(Select)>
  </ForwardRef(FormControl)>
  <ForwardRef(Button)
    className="makeStyles-filterButton-4"
    color="primary"
    fullWidth={false}
    id="filterButton"
    onClick={[Function]}
    type="submit"
    variant="contained"
  >
    Filter
  </ForwardRef(Button)>
</form>
`;
